Kong ingress controller(v3.3) panic: runtime error: invalid memory address or nil pointer dereference [recovered]

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

Getting panic error:

` >
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
    panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1d50bd2]

goroutine 527 [running]:
k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x26f0068, 0x39fa480}, {0x2028900, 0x399b6e0}, {0x39fa480, 0x0, 0x100000000000000?})
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/runtime/runtime.go:89 +0xee
k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0x0?})
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/runtime/runtime.go:59 +0x108
panic({0x2028900?, 0x399b6e0?})
    /usr/local/go/src/runtime/panic.go:785 +0x132
github.com/kong/kubernetes-ingress-controller/v3/internal/controllers/configuration.(*KongUpstreamPolicyReconciler).getUpstreamPoliciesForIngressServices(0xc000029730, {0x26f02d0, 0xc0008289b0}, {0x271ba90?, 0xc0008634a0})
    /workspace/internal/controllers/configuration/kongupstreampolicy_controller.go:327 +0xb2
sigs.k8s.io/controller-runtime/pkg/handler.(*enqueueRequestsFromMapFunc[...]).mapAndEnqueue(0x2717080?, {0x26f02d0, 0xc0008289b0?}, {0x27008e0?, 0xc000772d60}, {0x271ba90, 0xc0008634a0?}, 0xc0007cfcd0?)
    /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/handler/enqueue_mapped.go:119 +0x67
sigs.k8s.io/controller-runtime/pkg/handler.(*enqueueRequestsFromMapFunc[...]).Create(...)
    /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/handler/enqueue_mapped.go:84
sigs.k8s.io/controller-runtime/pkg/internal/source.(*EventHandler[...]).OnAdd(0x14, {0x2361b60?, 0xc0008634a0})
    /home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.19.0/pkg/internal/source/event_handler.go:92 +0x255
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnAdd(...)
    /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/controller.go:246
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
    /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/shared_informer.go:978 +0x139
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/wait/backoff.go:226 +0x33
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0007cff70, {0x26d1240, 0xc000edf260}, 0x1, 0xc000ef5030)
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/wait/backoff.go:227 +0xaf
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000ef8770, 0x3b9aca00, 0x0, 0x1, 0xc000ef5030)
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/wait/backoff.go:204 +0x7f
k8s.io/apimachinery/pkg/util/wait.Until(...)
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/wait/backoff.go:161
k8s.io/client-go/tools/cache.(*processorListener).run(0xc000fffa70)
    /home/runner/go/pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/shared_informer.go:972 +0x5a
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/wait/wait.go:72 +0x4c
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 501
    /home/runner/go/pkg/mod/k8s.io/apimachinery@v0.31.0/pkg/util/wait/wait.go:70 +0x73`

### Expected Behavior

Kong ingress controller pod should come up.

### Steps To Reproduce

```markdown
I tried with below commands, but result is same.

method:1
helm install kong kong/ingress -n kong --create-namespace 

method:2
helm upgrade --install controller kong/kong -n kic --create-namespace --set ingressController.enabled=true --set ingressController.gatewayDiscovery.enabled=true --set ingressController.gatewayDiscovery.adminApiService.name=gateway-kong-admin --set deployment.kong.enabled=false --set replicaCount=2
```


### Kong Ingress Controller version

```shell
3.3
```


### Kubernetes version

```shell
1.30.5-gke.1014001
```


### Anything else?

I followed the below documentation, but not working.

https://docs.konghq.com/kubernetes-ingress-controller/latest/install/helm/

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -110,7 +110,9 @@ Adding a new version? You'll need three changes:
 - Fixed annotation `konghq.com/rewrite` that was not being applied sometimes
   when `Ingress` without annotation and a different `Ingress` with annotation
   pointed to the same `Service`.
-  [#6569](https://github.com/Kong/kubernetes-ingress-controller/pull/6626)
+  [#6626](https://github.com/Kong/kubernetes-ingress-controller/pull/6626)
+- Fix panic in `KongUpstreamPolicyReconciler` when using with Ingress having a nil HTTP rule
+  [#6651](https://github.com/Kong/kubernetes-ingress-controller/pull/6651)
 
 ## [3.3.1]
 

--- a/internal/controllers/configuration/kongupstreampolicy_controller.go
+++ b/internal/controllers/configuration/kongupstreampolicy_controller.go
@@ -324,18 +324,25 @@ func (r *KongUpstreamPolicyReconciler) getUpstreamPoliciesForIngressServices(ctx
 	}
 	var requests []reconcile.Request
 	for _, rule := range ingress.Spec.Rules {
+		if rule.HTTP == nil {
+			continue
+		}
+
 		for _, path := range rule.HTTP.Paths {
 			if path.Backend.Service == nil {
 				continue
 			}
-			service := &corev1.Service{}
-			if err := r.Client.Get(ctx, k8stypes.NamespacedName{
-				Namespace: ingress.Namespace,
-				Name:      path.Backend.Service.Name,
-			}, service); err != nil {
+			var (
+				nn = k8stypes.NamespacedName{
+					Namespace: ingress.Namespace,
+					Name:      path.Backend.Service.Name,
+				}
+				service corev1.Service
+			)
+			if err := r.Client.Get(ctx, nn, &service); err != nil {
 				if !apierrors.IsNotFound(err) {
 					r.Log.Error(err, "Failed to retrieve Service in watch predicates",
-						"Service", fmt.Sprintf("%s/%s", ingress.Namespace, path.Backend.Service.Name),
+						"Service", nn,
 					)
 				}
 				continue

--- a/internal/controllers/configuration/kongupstreampolicy_controller_test.go
+++ b/internal/controllers/configuration/kongupstreampolicy_controller_test.go
@@ -1,16 +1,23 @@
 package configuration
 
 import (
+	"context"
 	"testing"
 
+	"github.com/go-logr/logr"
 	"github.com/samber/lo"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	corev1 "k8s.io/api/core/v1"
 	netv1 "k8s.io/api/networking/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
 
 	"github.com/kong/kubernetes-ingress-controller/v3/internal/gatewayapi"
+	kongv1beta1 "github.com/kong/kubernetes-ingress-controller/v3/pkg/apis/configuration/v1beta1"
 )
 
 func TestIndexIngressesOnBackendServiceName(t *testing.T) {
@@ -179,6 +186,237 @@ func TestIndexRoutesOnBackendRefServiceFacadeName(t *testing.T) {
 		t.Run(tc.name, func(t *testing.T) {
 			indexes := indexRoutesOnBackendRefServiceFacadeName(tc.object)
 			require.ElementsMatch(t, tc.expectedIndexes, indexes, "Should return expected indexes")
+		})
+	}
+}
+
+func TestGetUpstreamPoliciesForIngressServices(t *testing.T) {
+	tests := []struct {
+		name             string
+		services         []client.Object
+		ingress          netv1.Ingress
+		expectedRequests []ctrl.Request
+	}{
+		{
+			name: "Multiple services with annotations",
+			services: []client.Object{
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service1",
+						Namespace: "default",
+						Annotations: map[string]string{
+							kongv1beta1.KongUpstreamPolicyAnnotationKey: "policy1",
+						},
+					},
+				},
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service2",
+						Namespace: "default",
+						Annotations: map[string]string{
+							kongv1beta1.KongUpstreamPolicyAnnotationKey: "policy2",
+						},
+					},
+				},
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service3",
+						Namespace: "default",
+					},
+				},
+			},
+			ingress: netv1.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-ingress",
+					Namespace: "default",
+				},
+				Spec: netv1.IngressSpec{
+					Rules: []netv1.IngressRule{
+						{
+							IngressRuleValue: netv1.IngressRuleValue{
+								HTTP: &netv1.HTTPIngressRuleValue{
+									Paths: []netv1.HTTPIngressPath{
+										{
+											Backend: netv1.IngressBackend{
+												Service: &netv1.IngressServiceBackend{
+													Name: "service1",
+												},
+											},
+										},
+										{
+											Backend: netv1.IngressBackend{
+												Service: &netv1.IngressServiceBackend{
+													Name: "service2",
+												},
+											},
+										},
+										{
+											Backend: netv1.IngressBackend{
+												Service: &netv1.IngressServiceBackend{
+													Name: "service3",
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedRequests: []ctrl.Request{
+				{
+					NamespacedName: k8stypes.NamespacedName{
+						Namespace: "default", Name: "policy1",
+					},
+				},
+				{
+					NamespacedName: k8stypes.NamespacedName{
+						Namespace: "default", Name: "policy2",
+					},
+				},
+			},
+		},
+		{
+			name: "No services with annotations",
+			services: []client.Object{
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service1",
+						Namespace: "default",
+					},
+				},
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service2",
+						Namespace: "default",
+					},
+				},
+			},
+			ingress: netv1.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-ingress",
+					Namespace: "default",
+				},
+				Spec: netv1.IngressSpec{
+					Rules: []netv1.IngressRule{
+						{
+							IngressRuleValue: netv1.IngressRuleValue{
+								HTTP: &netv1.HTTPIngressRuleValue{
+									Paths: []netv1.HTTPIngressPath{
+										{
+											Backend: netv1.IngressBackend{
+												Service: &netv1.IngressServiceBackend{
+													Name: "service1",
+												},
+											},
+										},
+										{
+											Backend: netv1.IngressBackend{
+												Service: &netv1.IngressServiceBackend{
+													Name: "service2",
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedRequests: []ctrl.Request{},
+		},
+		{
+			name: "Service in different namespace cannot be used",
+			services: []client.Object{
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service1",
+						Namespace: "other",
+						Annotations: map[string]string{
+							kongv1beta1.KongUpstreamPolicyAnnotationKey: "policy1",
+						},
+					},
+				},
+			},
+			ingress: netv1.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-ingress",
+					Namespace: "default",
+				},
+				Spec: netv1.IngressSpec{
+					Rules: []netv1.IngressRule{
+						{
+							IngressRuleValue: netv1.IngressRuleValue{
+								HTTP: &netv1.HTTPIngressRuleValue{
+									Paths: []netv1.HTTPIngressPath{
+										{
+											Backend: netv1.IngressBackend{
+												Service: &netv1.IngressServiceBackend{
+													Name: "service1",
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			expectedRequests: []ctrl.Request{},
+		},
+		{
+			name: "Ingress without HTTP rules does not panic",
+			services: []client.Object{
+				&corev1.Service{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "service1",
+						Namespace: "default",
+						Annotations: map[string]string{
+							kongv1beta1.KongUpstreamPolicyAnnotationKey: "policy1",
+						},
+					},
+				},
+			},
+			ingress: netv1.Ingress{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-ingress",
+					Namespace: "default",
+				},
+				Spec: netv1.IngressSpec{
+					Rules: []netv1.IngressRule{
+						{
+							IngressRuleValue: netv1.IngressRuleValue{
+								HTTP: nil,
+							},
+						},
+					},
+				},
+			},
+			expectedRequests: []ctrl.Request{},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// Create a fake client
+			fakeClient := fake.NewClientBuilder().
+				WithObjects(tt.services...).
+				Build()
+
+			// Create a KongUpstreamPolicyReconciler
+			reconciler := &KongUpstreamPolicyReconciler{
+				Client: fakeClient,
+				Log:    logr.Discard(),
+			}
+
+			// Call the function
+			requests := reconciler.getUpstreamPoliciesForIngressServices(context.Background(), &tt.ingress)
+
+			// Assert the results
+			assert.ElementsMatch(t, tt.expectedRequests, requests)
 		})
 	}
 }

